libcodechecker/analyze/tu_collector.py

```diff
@@ -0,0 +1,257 @@
+            dependencies |= set(__gather_dependencies(command, build_dir))
+        except Exception as ex:
+            logging.debug("Couldn't create dependencies:")
+            logging.debug(str(ex))
+            error += str(ex)
+
+    return dependencies, error
+
+
+def add_sources_to_zip(zip_file, files):
+    """
+    This function adds source files to the ZIP file if those are not present
+    yet. The files will be placed to the "sources-root" directory under the ZIP
+    file.
+
+    zip_file -- A ZIP file name as a string.
+    files -- A file path or an iterable of file paths.
+    """
+    if isinstance(files, str):
+        files = [files]
+
+    with zipfile.ZipFile(zip_file, 'a') as archive:
+        for f in files:
+            archive_path = os.path.join('sources-root', f.lstrip('/'))
+
+            try:
+                archive.getinfo(archive_path)
+            except KeyError:
+                archive.write(f, archive_path, zipfile.ZIP_DEFLATED)
+            else:
+                logging.debug("'{}' is already in the ZIP file, won't add it "
+                              "again!".format(f))
+
+
+def zip_tu_files(zip_file, compilation_database, write_mode='w'):
+    """
+    Collects all files to a zip file which are required for the compilation of
+    the translation units described by the given compilation database.
+    If there are some files which couldn't be collected then a "no-sources"
+    file will contain the error message of its reason.
+    The function returns the set of files.
+
+    zip_file -- A file name or a file object.
+    compilation_database -- Either a path of the compilation database JSON file
+                            or a list of the parsed JSON.
+    write_mode -- The file opening mode of the zip_file. In case of 'a' the new
+                  files are appended to the existing zip file, in case of 'w'
+                  the files are added to a clean zip file.
+    """
+    if isinstance(compilation_database, str) or \
+            isinstance(compilation_database, unicode):
+        with open(compilation_database) as f:
+            compilation_database = json.load(f)
+
+    no_sources = 'no-sources'
+    tu_files = set()
+    error_messages = ''
+
+    for buildaction in compilation_database:
+        files, err = get_dependent_headers(
+            buildaction['command'],
+            buildaction['directory'])
+
+        tu_files |= files
+
+        if err:
+            error_messages += buildaction['file'] + '\n' \
+                + '-' * len(buildaction['file']) + '\n' + err + '\n'
+
+    if write_mode == 'a' and os.path.isfile(zip_file):
+        with zipfile.ZipFile(zip_file) as archive:
+            try:
+                archive.getinfo(no_sources)
+            except KeyError:
+                pass
+            else:
+                no_sources = 'no-sources_' + __random_string(5)
+
+    with zipfile.ZipFile(zip_file, write_mode) as archive:
```

Comment:
If `error_messages` variable is empty we do not write any content to the zip file. So we can check first, than write content to the zip:
```py
if error_messages:
    with zipfile.ZipFile(zip_file, write_mode) as archive:
        archive.writestr(no_sources, error_messages)
```